### ISSUE TITLE
HTTP client - option to bind to source address for sending request

### ISSUE THREAD

**Issue by jeyavel** [annoviko/robotframework-httpctrl#21]
**Is your feature request related to a problem? Please describe.**
When you are using a device with multiple network interfaces, allow option to bind the interface for sending HTTP/REST request.
Missing the source interface binding option becomes Blocker in scenarios where it is required to send/receive requests/responses on specific interfaces.

**Describe the solution you'd like**
python http.client library used as base for this HTTPCtrl already supports this source address binding. Utilise the option and make source_address as optional parameter.
Changes I made and tested for my learning:

source_address needs to be a tuple of ip_address and port number. Can include valdiation for ipv4/ipv6 address and valid port number - similar to destination ip address and port number we already have.

1. HTTPCtrl-->__init__.py 
     a. def __init__ to include self.__source.address=source.address
     b. def initialize_client to include a parameter source_address=None, source_port=0 and  self.__source_address assignment when it has value. 
     c. Inside def __send function - when creating http.client.HTTPConnection, pass on the source_address parameter
 
**Describe alternatives you've considered**
In Robot Framework there is no alternate directly available. Only option is to customize the existing keyword HTTPCtrl or Request Library or use python library.

**Additional context**
--None--

**Comment by annoviko** [comment]
Hello @jeyavel ,

Thank you for the request. I can introduce the functionality, but I need your help with testing when a machine has multiple network interfaces (I cannot test it on my machine, mine has only one. I can only check that client binds to proper interface and can be bind to the specific address/port).

Not sure if it is possible in python, but 10-15 years ago I had a problem with multiple network interfaces and it was needed to specify the concrete network interface, like `eth0`, `eth1`, etc. Not sure if it is needed for you as well.

**Comment by jeyavel** [comment]
Hi @annoviko,Apppreciate your quick follow-up.Definitely I can help you in the testing.The http.client used for creating this library is already having this facility to bind to source address.I personally checked with ip address and port 0 (any port).Not sure of using interface name option support. If you think about interface name option, will it recognise and bind with virtual interfaces. Example eth0:1, eth0:2 Please let me know when build is available and how to take the build for my installation. I will test and return back with my observations.Best Regards,JeyavelOn Sep 24, 2021 20:54, Andrei Novikov ***@***.***> wrote:
Hello @jeyavel ,
Thank you for the request. I can introduce the functionality, but I need your help with testing when a machine has multiple network interfaces (I cannot test it on my machine, mine has only one. I can only check that client binds to proper interface and can be bind to the specific address/port).
Not sure if it is possible in python, but 10-15 years ago I had a problem with multiple network interfaces and it was needed to specify the concrete network interface, like eth0, eth1, etc. Not sure if it is needed for you as well.

**Comment by annoviko** [comment]
Hello @jeyavel 

I released version 0.1.7 (available on pip): https://pypi.org/project/robotframework-httpctrl/0.1.7/
```
pip3 install robotframework-httpctrl
```

There is a usage example:
```
Initialize Client   192.168.0.5   8000   192.168.0.1   8001
```

I tested in by sending request from 127.0.0.1:8001 to sever 127.0.0.1:8000 and check if client has proper address and port:
```
Bind Client to Address
    [Teardown]  Stop Server
    Initialize Client   127.0.0.1   8000   127.0.0.1   8001
    Start Server        127.0.0.1   8000

    ${connection 1}=   Send HTTP Request Async   POST   /post   Post Message

    ${response}=   Get Async Response   ${connection 1}   1
    Should Be Equal   ${response}   ${None}

    Wait For Request

    ${source address}=   Get Request Source Address
    ${source port}=      Get Request Source Port
    ${source port as integer}=   Get Request Source Port As Integer

    Should Be Equal   ${source address}   127.0.0.1
    Should Be Equal   ${source port}      8001
    Should Be Equal   ${source port as integer}   ${8001}

    Reply By   200   Post Response

    ${response}=   Get Async Response   ${connection 1}   1
    Should Not Be Equal   ${response}   ${None}
```

**Comment by jeyavel** [comment]
Thank you @annoviko 
I will use this new build and get back with you on my test status.

**Comment by jeyavel** [comment]
When client port is not provided - currently it returns 

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'

In the code it is related to 
def __get_source_address(self):
        if (self.__client_host is None) and (self.__client_port is None):
            return None
        return self.__client_host, int(self.__client_port)

I would recommend to keep client_port as 0 - when not available. This will make the socket listening on the given ip at any dynamically assigned port.

**Comment by annoviko** [comment]
Hello @jeyavel ,

Make sense, client port is optional now (port is going to be dynamically assigned):
```
Initialize Client   192.168.0.5   8000   192.168.0.1
```

The update available in 0.1.8: https://pypi.org/project/robotframework-httpctrl/0.1.8/
```
pip3 install robotframework-httpctrl
```
